src/java.base/share/classes/java/lang/foreign/Arena.java

```diff
@@ -0,0 +1,154 @@
+ * arena are associated with the arena scope. This makes arenas extremely useful when interacting with foreign code, as shown below:
+ *
+ * {@snippet lang = java:
+ * try (Arena arena = Arena.openConfined()) {
+ *     MemorySegment nativeArray = arena.allocateArray(ValueLayout.JAVA_INT, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+ *     MemorySegment nativeString = arena.allocateUtf8String("Hello!");
+ *     MemorySegment upcallStub = linker.upcallStub(handle, desc, arena.scope());
+ *     ...
+ * } // memory released here
+ *}
+ *
+ * <h2 id = "thread-confinement">Safety and thread-confinement</h2>
+ *
+ * Arenas provide strong temporal safety guarantees: a memory segment allocated by an arena cannot be accessed
+ * <em>after</em> the arena has been closed. The cost of providing this guarantee varies based on the
+ * number of threads that have access to the memory segments allocated by the arena. For instance, if an arena
+ * is always created and closed by one thread, and the memory segments associated with the arena's scope are always
+ * accessed by that same thread, then ensuring correctness is trivial.
+ * <p>
+ * Conversely, if an arena allocates segments that can be accessed by multiple threads, or if the arena can be closed
+ * by a thread other than the accessing thread, then ensuring correctness is much more complex. For example, a segment
+ * allocated with the arena might be accessed <em>while</em> another thread attempts, concurrently, to close the arena.
+ * To provide the strong temporal safety guarantee without forcing every client, even simple ones, to incur a performance
+ * impact, arenas are divided into <em>thread-confined</em> arenas, and <em>shared</em> arenas.
+ * <p>
+ * Confined arenas, support strong thread-confinement guarantees. Upon creation, they are assigned an
+ * {@linkplain #isCloseableBy(Thread) owner thread}, typically the thread which initiated the creation operation.
+ * The segments created by a confined arena can only be {@linkplain SegmentScope#isAccessibleBy(Thread) accessed}
+ * by the owner thread. Moreover, any attempt to close the confined arena from a thread other than the owner thread will
+ * fail with {@link WrongThreadException}.
+ * <p>
+ * Shared arenas, on the other hand, have no owner thread. The segments created by a shared arena
+ * can be {@linkplain SegmentScope#isAccessibleBy(Thread) accessed} by multiple threads. This might be useful when
+ * multiple threads need to access the same memory segment concurrently (e.g. in the case of parallel processing).
+ * Moreover, a shared arena {@linkplain #isCloseableBy(Thread) can be closed} by any thread.
+ *
+ * @since 20
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+public interface Arena extends SegmentAllocator, AutoCloseable {
+
+    /**
+     * Returns a native memory segment with the given size (in bytes) and alignment constraint (in bytes).
```

Comment:
It is noted that the current documentation does not require a **new** native memory segment to be returned. Would it not be better with:

Creates a new native memory segment ...

The new shared segment might share actual backing memory though.

Reply:
My feeling is that being overly precise over identity might backfire. It is not important whether the segment is a new instance or not. But there is, perhaps, another invariant that is more semantically relevant: e.g. the returned segments (whether new or not, we don't care) should be backed by "disjoint" regions of memory. That is, if the method returns a segment with address `0` and size `100`, calling the method again cannot return a segment whose address is `50` and size is `100`. In principle, the segment allocator interface allows for this (see `SegmentAllocator::prefixAllocator`) - but for an arena, a behavior such as this would be indesirable, IMHO.

Reply:
I will add:

```
Furthermore, for any two segments S1, S2 returned by this method, the following invariant must hold:

 S1.overlappingSlice(S2).isEmpty() == true
 ```